File: algorithm/java/src/main/java/com/liyongquan/interview/FindBigger.java
package com.liyongquan.interview;

import java.util.Deque;
import java.util.LinkedList;

/**
 * 找到右边第一个大的值
 */
public class FindBigger {
    public int[] find(int[] arr) {
        int len = arr.length;
        Deque<Integer> pq = new LinkedList<>();
        int[] res = new int[len];
        for (int i = len - 1; i >= 0; i--) {
            while (pq.size() > 0 && pq.peekFirst() < arr[i]) {
                pq.pollFirst();
            }
            res[i] = pq.size() == 0 ? -1 : pq.peekFirst();
            pq.offerFirst(arr[i]);
        }
        return res;
    }
}

File: algorithm/golang/interview/FindBigger.go
package interview

//字节面试题
//查找右边第一个比自身大的数字
func findBigger(arr []int) []int {
	stack := make([]int, 0)
	res := make([]int, len(arr))
	for i := len(arr) - 1; i >= 0; i-- {
		//单调栈
		for len(stack) > 0 && stack[len(stack)-1] < arr[i] {
			stack = stack[0 : len(stack)-1]
		}
		if len(stack) > 0 {
			res[i] = stack[len(stack)-1]
		} else {
			res[i] = -1
		}
		stack = append(stack, arr[i])
	}
	return res
}
